From 77d52338b1cb4cf9f215310357f049b6e6bd2231 Mon Sep 17 00:00:00 2001
From: Jiping Ma <jiping.ma2@windriver.com>
Date: Sun, 17 Sep 2023 18:13:46 -0700
Subject: [PATCH 2/3] Do not load unnecessary device drivers in the kdump
 initramfs

Make vmcore collection more reliable

This patch makes kernel crash dump (vmcore) collection more reliable by making
the following changes to the kdump configuration:

- We don't need pci AER functionality enabled in the kdump kernel,
  and it causes some boards to hang in certain situations. So just
  disable it.

- systemd.mask=systemd-sysctl.service kernel command line argument is utilized
  to avoid applying sysctl settings in kdump sessions. One problematic sysctl
  is vm.min_free_kbytes, which, on certain StarlingX deployments, is set to
  1179648 KiB (more than 1 GiB) via "/etc/sysctl.conf". This causes the kdump
  kernel to start the out-of-memory task killing procedure right after systemd
  applies the sysctl settings, at least in some system configurations. (Recall
  that StarlingX reserves 2 GiB for the kdump kernel by default.)

- The udev.children_max=2 kernel command line argument is utilized to prevent
  the udev daemon from spawning many child tasks concurrently, in an effort to
  reduce instantaneous memory utilization in kdump sessions.

- The dependency of the kdump-tools-dump.service unit on the
  network-online.target unit is removed and the argument
  systemd.mask=networking.service is added to the kernel command line, which
  collectively let us prevent the networking.service from starting in kdump
  sessions, which in turn avoids unnecessarily attempting to bring up network
  interfaces.

- The nousb kernel command line argument appears to have been deprecated, so
  it is replaced with the usbcore.nousb kernel command line argument, which
  prevents the initialization of the USB subsystem in kdump sessions as
  expected.

- The following modules are prevented from being loaded in kdump sessions,
  which reduces memory consumption: "ice,i40e,iavf,bnxt_en,bnxt_re,ib_core,
  mlx5_core,mlx5_ib,mlx_compat,vfio,oct_ep_phc,e1000e,ixgbe,ixgbevf,drbd,
  tg3,igb". This is done via the use of the module_blacklist= kernel command
  line argument.

- cgroup_disable=memory is utilized to disable the memory cgroups, which is
  intended for reducing memory utilization. This was also mentioned at:
  https://bugzilla.redhat.com/show_bug.cgi?id=605717

- transparent_hugepage=never is utilized to avoid enabling transparent
  hugepages unintendedly in kdump sessions. This was done out of caution,
  given that the kdump session in Debian-based StarlingX pivot_root's into the
  regular root file system and starts almost like a regular system.

- Finally, the systemd.mask=iscsid.service argument is utilized to avoid
  starting the iSCSI daemon in kdump session, which we believe to be
  unnecessary in a vmcore collection scenario.

Signed-off-by: M. Vefa Bicakci <vefa.bicakci@windriver.com>
Signed-off-by: Jiping Ma <jiping.ma2@windriver.com>
Signed-off-by: Catalin Rapiteanu <viorel-catalin.rapiteanu@windriver.com>
---
 debian/rules | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/debian/rules b/debian/rules
index 537430e..1db33eb 100755
--- a/debian/rules
+++ b/debian/rules
@@ -14,7 +14,10 @@ ifeq ($(DEB_HOST_ARCH),arm64)
 else ifeq ($(DEB_HOST_ARCH),ppc64el)
 	KDUMP_CMDLINE_APPEND += maxcpus=1 irqpoll noirqdistrib usbcore.nousb
 else
-	KDUMP_CMDLINE_APPEND += nr_cpus=1 irqpoll usbcore.nousb
+	KDUMP_CMDLINE_APPEND += nr_cpus=1 irqpoll usbcore.nousb pci=noaer
+	KDUMP_CMDLINE_APPEND += module_blacklist=ice,i40e,iavf,bnxt_en,bnxt_re,ib_core,mlx5_core,mlx5_ib,mlx_compat,vfio,oct_ep_phc,e1000e,ixgbe,ixgbevf,drbd,tg3,igb
+	KDUMP_CMDLINE_APPEND += cgroup_disable=memory transparent_hugepage=never systemd.mask=systemd-sysctl.service systemd.mask=networking.service
+	KDUMP_CMDLINE_APPEND += systemd.mask=iscsid.service udev.children_max=2
 endif
 
 %:
-- 
2.47.3

